Error message truncation

matajoh opened the pull request:
This PR adds the ability to set a maximum error message length for a policy enforcer.

anmaxvl commented on internal/guest/runtime/hcsv2/uvm.go:
```diff
@@ -107,13 +107,18 @@ func (h *Host) SetConfidentialUVMOptions(ctx context.Context, r *guestresource.L
 
+	// this limit ensures messages are below the truncation limit that
+	// is imposed by Service Fabric
+	maxErrorMessageLength := 3 * 1024
```
any reason it has to be a parameter, rather than a security policy package const or something? this way we don't have to update signatures.

matajoh replied:
Our tests depend on being able to read out values from the error messages to validate that the errors are being created correctly. If the max value is set to what it needs to be for SF, then you get random false negatives from the test suite. Having it be a parameter avoids that difficulty.

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -333,36 +325,176 @@ func (policy *regoEnforcer) denyWithError(ctx context.Context, policyError error
-	return fmt.Errorf(policyDecisionPattern, base64EncodedDecisionJSON)
+		decisionJSON, err := json.Marshal(decision)
+		if err != nil {
+			log.G(ctx).Errorf("Unable to marshal error object: %v", err)
```
`log.G(ctx).WithError(err).Error("unable to marshal error object")`

matajoh replied:
Gotcha.

anmaxvl commented on pkg/securitypolicy/securitypolicyenforcer_rego.go:
```diff
@@ -333,36 +325,176 @@ func (policy *regoEnforcer) denyWithError(ctx context.Context, policyError error
+			// check if it is a framework reason object
+			if reason, ok := rawReason.(rpi.RegoQueryResult); ok {
+				// check if we can remove error_objects
+				if _, ok := reason["error_objects"]; ok {
```
we'll always reach here regardless if we need to remove `error_objects` or not. anything else we could potentially want to truncate in the future? we could potentially define a "truncateFunc" and iterate/call the funcs in a loop or something instead of potentially adding more `if !actionTaken` statements. but it's  probably ok for now...

matajoh replied:
I like the idea of a truncate function, that's a cleaner way of doing this.

matajoh replied:
Latest version uses this approach, I like it a lot better

matajoh committed:
Policy decision truncation.

In some circumstances, the policy decision object returned from a policy
denial causes the resulting error message to exceed the maximum error length
imposed by Service Fabric. This PR adds some truncation logic to reduce the
size of the decision object so it first into the limit. Firstly, all standard
capability sets (privileged and unprivileged) are replaced with a placeholder.
Then, if the message is above the length limit then the following things are
truncated until the message is below the threshold:

1. `reason.error_objects`
2. `input`
3. `reason` (the rest of the reason object returned from the policy)

Signed-off-by: Matthew A Johnson <matjoh@microsoft.com>
